modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientPrimaryReplicaTracker.java

```diff
@@ -0,0 +1,253 @@
+                // TODO: Should we wait for table sync?
+                return CompletableFuture.failedFuture(tableNotFoundException(tableId));
+            }
+
+            CatalogZoneDescriptor zoneDesc = catalogService.zone(tableDesc.zoneId(), timestamp);
+
+            if (zoneDesc == null) {
+                return CompletableFuture.failedFuture(tableNotFoundException(tableId));
+            }
+
+            return primaryReplicasAsyncInternal(tableId, zoneDesc.partitions());
+        } catch (Throwable t) {
+            return CompletableFuture.failedFuture(t);
+        }
+    }
+
+    private CompletableFuture<List<ReplicaHolder>> primaryReplicasAsyncInternal(int tableId, int partitions) {
+        CompletableFuture<ReplicaMeta>[] futs = (CompletableFuture<ReplicaMeta>[]) new CompletableFuture[partitions];
+
+        for (int partition = 0; partition < partitions; partition++) {
+            futs[partition] = placementDriver.getPrimaryReplica(new TablePartitionId(tableId, partition), clock.now());
+        }
+
+        CompletableFuture<Void> all = CompletableFuture.allOf(futs);
+
+        return all.thenApply(v -> {
+            List<ReplicaHolder> replicaHolders = new ArrayList<>(partitions);
+
+            for (int partition = 0; partition < partitions; partition++) {
+                ReplicaMeta replicaMeta = futs[partition].join();
+
+                // Returning null is fine - the client will use default channel for this partition.
+                replicaHolders.add(replicaMeta == null
+                        ? new ReplicaHolder(null, HybridTimestamp.MIN_VALUE)
+                        : new ReplicaHolder(replicaMeta.getLeaseholder(), replicaMeta.getStartTime()));
+            }
+
+            return replicaHolders;
+        });
+    }
+
+    @Override
+    public CompletableFuture<Boolean> notify(EventParameters parameters, @Nullable Throwable exception) {
+        if (exception != null) {
+            return CompletableFuture.completedFuture(false);
+        }
+
+        if (parameters instanceof DropTableEventParameters) {
+            DropTableEventParameters dropTableEvent = (DropTableEventParameters) parameters;
+            primaryReplicas.remove(dropTableEvent.tableId());
+
+            return CompletableFuture.completedFuture(false);
+        }
+
+        if (!(parameters instanceof PrimaryReplicaEventParameters)) {
+            assert false : "Unexpected event parameters: " + parameters.getClass();
+
+            return CompletableFuture.completedFuture(false);
+        }
+
+        PrimaryReplicaEventParameters primaryReplicaEvent = (PrimaryReplicaEventParameters) parameters;
+        if (!(primaryReplicaEvent.groupId() instanceof TablePartitionId)) {
+            return CompletableFuture.completedFuture(false);
+        }
+
+        TablePartitionId tablePartitionId = (TablePartitionId) primaryReplicaEvent.groupId();
+        var hld = primaryReplicas.get(tablePartitionId.tableId());
+
+        if (hld != null) {
```

Comment:
Seems that you have a race here.
0. On start you've registered a listener on primary replica election.
1. Thread 1 at some point fixes the getPrimaryTimestamp and hangs.
2. Primary expires and new one is elected - you receive the notification and ignore it
3. Thread 1 wakes up and populates the map with obsolete primary replica.

I believe that you can use primaryReplica startTime in order to linearise all possible updates and thus understand whether you need to update corresponding partition meta or not.

Reply:
Fixed - `updatePrimaryReplica` now uses `leaseStartTime` to skip stale updates.